### PR TITLE
Galaxy Workflow support for GenomeSpace

### PR DESCRIPTION
**Pull request by nuwang**
This pull request adds workflow integration for GenomeSpace import/export tools. It's been a long time coming since the original mailing list post almost 2 years ago: http://dev.list.galaxyproject.org/Genomespace-Integration-with-workflows-td4664513.html. I also talked to @jmchilton during GCC 2014 about this - and for various reasons, it's been languishing in the backburner since.

This is a patch against the latest dev branch. We've coordinated with Ted Liefield (@liefield) from the GenomeSpace team to get some support functions added into GenomeSpace for obtaining an auth-token that can be used for performing the import/export. The open-id based method that @blankenberg added to the importer/exporter also works. There's a slight tension between security and ease-of-use, but Ted reckons that since the token is relatively short-lived and revokable, it doesn't pose a significant security threat. If there are any problems here, we can modify it again to mandate openid login, but at the cost of being less user-friendly. (the current version allows for direct import/export from GenomeSpace with only a GenomeSpace login)

The GenomeSpace importer/exporter itself has been rewritten as a standalone pip installable tool, available here: https://github.com/gvlproject/python-genomespaceclient. We hope to transfer that code back back into GenomeSpace or Galaxy as a set of Python bindings + commandline client for GenomeSpace.

There's a 3 minute video of how things work here:
https://www.youtube.com/watch?v=5QPtWS_ab0I

Feedback appreciated.

### CODE REVIEW

File: client/galaxy/scripts/mvc/tool/tool-genomespace.js
Comment (nuwang) [inline client/galaxy/scripts/mvc/tool/tool-genomespace.js:2]: Not sure whether this folder is the best location for this support module.

File: client/galaxy/scripts/mvc/tool/tool-genomespace.js
Comment (nuwang) [inline client/galaxy/scripts/mvc/tool/tool-genomespace.js]: Could this location be made centrally configurable somehow?  If so, that would be pretty useful since it could be made configurable at install time to point to the Australian mirror for example.

**Comment by bgruening** [comment]
This is awesome! Great work @nuwang!

File: lib/galaxy/web/form_builder.py
Comment (nuwang) [inline lib/galaxy/web/form_builder.py:220]: When we originally started working on this, the tool parameters were still being rendered on the server side. Since it's now being rendered on the client-side, we've added a backbone component for it, but also left this bit as is since workflows don't fully support client-side rendering yet. Once that conversion is done, I hope that this bit can be nuked, preferably from orbit :-)

**Comment by nuwang** [comment]
@bgruening Thanks! @ykowsar made the original modifications to GenomeSpace that enabled this to happen, and @liefield tested and mainlined the code.

**Comment by nuwang** [comment]
@blankenberg @martenson How's this looking? About 2 months ago, I made another set of fixes to make sure the merge could be done smoothly (all pylint warnings fixed, merge issues fixed) but it seems to have drifted out of sync again. It would be great if we can do something about this pull request because it's been languishing for quite a while now. Let me know if you'd like to see any more changes and any comments etc. I can make another update with the merge issues resolved.

**Comment by nuwang** [comment]
@blankenberg @martenson @afgane I've just made some commits updating the branch to latest dev and fixing all merge issues.

File: client/galaxy/scripts/mvc/tool/tool-genomespace.js
Comment (jmchilton) [reply client/galaxy/scripts/mvc/tool/tool-genomespace.js]: Galaxy.config may be available - that is like a global configuration object that you can populate with options from galaxy.ini and such. It is worth looking into.

**Comment by jmchilton** [comment]
I think we need to build the client before merging - do you mind if I do that and push to your branch at somepoint before merging the PR?

**Comment by nuwang** [comment]
@jmchilton Thanks, that'd be great. I didn't do that because I was afraid it would trigger merge conflicts more often. There's also a minor issue with compressed file handling. Simon and I thought this would fix that: https://github.com/galaxyproject/galaxy/pull/1814/commits/357f802c439b62a737da39950e0e2a719aa529c4 but the files remain stubbornly compressed. Is there a different way to trigger Galaxy's built in capabilities for this?

**Comment by nuwang** [comment]
@jmchilton We've identified the issue with compressed file handling and have logged an issue here: https://github.com/galaxyproject/galaxy/issues/4443
So we think the issue is fixed as far as this pull request goes.

I'll rebuild the client and push it in now.

**Comment by nuwang** [comment]
@jmchilton I've merged in the latest master, but haven't done a client rebuild, because that looks like something that should be done during the PR merge?

File: tools/genomespace/genomespace_importer.py
Comment (mvdbeek) [inline tools/genomespace/genomespace_importer.py]: If I understand this implementation correctly, all datatypes that are sniffed successfully will not be modified by this mapping, so I'd remove all `fasta/q*`, `gff`, `gtf`, `*txt` etc. types. In fact I'd probably eliminate this step, or, if you want to force things like `gtf`/`gff`, which are difficult to sniff and where you are certain that the genomespace extensions is correct, you'd have to change the order for this to occur before letting galaxy sniff the datatype.

Comment (mvdbeek) [reply tools/genomespace/genomespace_importer.py]: Also note that you can't force files to be of a datatype that they are not. Forcing fastqsager.gz to be fastqsanger will only change the datatype, not the content, so almost all tools (tools that discriminate between compressed and uncompressed files, and tools which will trigger an automatic conversion because they only support uncompressed) will fail on this dataset.

**Review by mvdbeek — COMMENTED** [review]
I'd keep compressed files compressed if galaxy can deal with it. I would think this becomes the default in the near future, but probably not for 17.09.

File: tools/genomespace/genomespace_importer.py
Comment (nuwang) [reply tools/genomespace/genomespace_importer.py]: @mvdbeek Thanks for reviewing. I've kept the original implementation logic as is, and have only refactored it (but have added more types). The specific line where filetype is determined is in the determine_file_type() method here: https://github.com/galaxyproject/galaxy/pull/1814/files/e44d67f1944aa43c9d28cd9efffca88cfecd51d5#diff-a9715b311879402b8b86a4013e4ec56bR152

The order of resolution is:
1. Respect genomespace supplied type
2. If no type, try to sniff type
3. Still no type, use filename extension
4. Still nothing, default to 'data'

There's no forced coercion unless the file is misnamed. Does that sound ok?

However, one thing I did change was to call sniff.handle_uploaded_dataset() to trigger decompression here prior to determining file type: https://github.com/galaxyproject/galaxy/pull/1814/files/e44d67f1944aa43c9d28cd9efffca88cfecd51d5#diff-a9715b311879402b8b86a4013e4ec56bR208

Given the current implementation of sniff.handle_uploaded_dataset(), only fasta.gz files will be decompressed. Everything else will be left as is.

Comment (mvdbeek) [reply tools/genomespace/genomespace_importer.py]: Sorry, I don't know how I got the order wrong, so that solves the one part, but you should remove the extensions that are not problematic in galaxy IMO.

So what you're saying is that only `fasta.gz` will be uncompressed ? Why only fasta.gz ? Or is that a typo and you meant fastq.gz ? Or is that because fasta is an example of a non-compressed datatype ? Normally compressed datatypes like fastq.gz should not be uncompressed by `handle_uploaded_dataset`.

Now the important thing is that if you respect the genomespace supplied type, which maps compressed fastq (`fqz`) to fastqsanger, you must uncompress the file. But it would be very good if that doesn't happen and you would not special-case compressed or uncompressed fastqs.

Comment (nuwang) [reply tools/genomespace/genomespace_importer.py]: @mvdbeek You're right - I missed that completely - thanks for catching that. I'll remove these newly added types.

File: tools/genomespace/genomespace_importer.py
Comment (mvdbeek) [inline tools/genomespace/genomespace_importer.py]: To clarify I think it would be good to change this to say that compressed files for non-compressed datatypes will be unpacked. I think that was part of the confusion.

Comment (nuwang) [reply tools/genomespace/genomespace_importer.py]: I've changed the wording as suggested.

File: tools/genomespace/genomespace_importer.py
Comment (nuwang) [reply tools/genomespace/genomespace_importer.py]: @mvdbeek This has been fixed now.

File: client/galaxy/scripts/mvc/tool/tool-genomespace.js
Comment (nuwang) [reply client/galaxy/scripts/mvc/tool/tool-genomespace.js]: @jmchilton Thanks. I've added a galaxy.ini config setting for the GenomeSpace URL now: https://github.com/galaxyproject/galaxy/pull/1814/commits/04a7e7676386b4818796822e24f648ded6c29d67

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: Should this be `Click the Browse button to select a file to export` ?

Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: Thanks for catching that! Fixed in: https://github.com/galaxyproject/galaxy/pull/1814/commits/efc078cff1c6dd58462ec1a6f74bb38da83f0483

**Comment by mvdbeek** [comment]
@nuwang can you rebase this against dev ? We have changed whitespaces all over the codebase (no whitespaces before/after parens anymore).

**Comment by nuwang** [comment]
@mvdbeek I've done that now, and have also squashed/fixed up commits so that the overall history is cleaner.

**Comment by mvdbeek** [comment]
So how do I setup a galaxy instance to use this tool ?
I have uncommented `openid_config_file = config/openid_conf.xml`, copied the sample file to `config/openid_conf.xml`. `openid_conf.xml`references a few additional files that do not ship with galaxy. So how can I get those / do I need these to run the tool ?

**Comment by mvdbeek** [comment]
Alright, should have just tried, now I have associated a Genomespace account with my galaxy user.
But the UI looks a little strange
screen shot 2017-08-18 at 12 34 36 pm

I'm supposed to be able to select files there, right ?

**Comment by nuwang** [comment]
Yes. You have to run "make client" to rebuild the client js files. I haven't done this as it triggers too many merge conflicts, and it should probably only be done prior to merge.

**Comment by mvdbeek** [comment]
If you're worried about merge conflicts you could add the static files in a separate commit. That way you just delete the last commit and build again.

**Comment by mvdbeek** [comment]
Alright, this works like a charm (I tried fastq, fastq.gz, txt), and I'd be happy to merge.
One little thing is that the alignment of the boxes in the tool-form isn't great:
screen shot 2017-08-18 at 12 46 09 pm

That's minor though. If you rebuild the client I will merge.

**Comment by mvdbeek** [comment]
I just noticed that the exporter let's you select a file in the genomespace browser, when you should be selecting a folder. That would also be cool to address, but the error message is clear, so that's OK for now.

**Comment by nuwang** [comment]
@mvdbeek Hmm... In the original video I posted, the browse button was on the left, and the two boxes were aligned correctly on the right. The line here used to do the trick: https://github.com/galaxyproject/galaxy/pull/1814/files#diff-1415f5af50ae7c6c205f1da7d8292e22R19
Something must have changed in Galaxy that's causing it to jump to the right?

**Comment by nuwang** [comment]
I've fixed the alignment in this commit: https://github.com/galaxyproject/galaxy/pull/1814/commits/40645f157ed42274a90e84a6c4d58b38a774d6a4
Now it looks like this:
image

**Comment by nuwang** [comment]
Many thanks for reviewing this Marius. I've just rebased on master, and rebuilt the client: https://github.com/galaxyproject/galaxy/pull/1814/commits/b6b4deb5a65f5f16c1c566a6391cc180f5964bff. It should be good to go.

WRT to restricting the exporter to select folders only - I think that would make sense. It requires changes in the GenomeSpace file browse dialogue though, which is internal to GenomeSpace, which is why it couldn't be done thus far. I'll file an issue on their end.

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: <command interpreter="python">genomespace_exporter.py
This is deprecated, can you change this to
```
 <command> python $__tool_directory__/genomespace_exporter.py
```
?

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: Maybe you'd want to replace `$input1.name` with `$input1.element_identifier` ? For normal datasets this falls back to `.name`, but usually for collection elements you want the element identifier.

File: tools/genomespace/genomespace_importer.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_importer.xml]: you probably want to replace this and the below line to `python $__tool_directory__/genomespace_importer.py`

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: The token worries me a little bit ... it will be logged and appears in the commandline.

**Review by mvdbeek — COMMENTED** [review]
Sorry for this piecemeal review, I didn't look at the xml carefully.

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: Can you bump the version? (0.0.5) seems fine.

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: can you place this in single quotes ?

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: Same here, everything that does not need to be expanded by the shell should be in single quotes

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: quotes

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_exporter.xml]: This doesn't do anything anymore.

File: tools/genomespace/genomespace_importer.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_importer.xml]: single quotes

File: tools/genomespace/genomespace_push.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_push.xml]: single quotes

File: tools/genomespace/genomespace_push.xml
Comment (mvdbeek) [inline tools/genomespace/genomespace_push.xml]: single quotes

File: tools/genomespace/genomespace_exporter.xml
Comment (mvdbeek) [reply tools/genomespace/genomespace_exporter.xml]: Maybe you could write this out to a file in cheetah and read it back in genomespace_exporter.py and genomespace_importer.py. That's not ideal either, but I guess that would be acceptable.

Comment (mvdbeek) [reply tools/genomespace/genomespace_exporter.xml]: `#open('token.txt', 'w').write($token)` should work, I would think.

Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: @mvdbeek Thanks, I'm attending to these now. Is there a way to make non data-source tools pass in a json_parameter_file with all params included (as in genomespace push here: https://github.com/galaxyproject/galaxy/pull/1814/files#diff-164cfd6b30a3542460531fab3994cfb5R10)? I found that only data-source tools do this. For non-data source tools, I had to override the exec_before_job hook to manually generate a json param dict: https://github.com/galaxyproject/galaxy/pull/1814/files#diff-a9715b311879402b8b86a4013e4ec56bR67

If there's a way to do this, it would significantly simplify the code, and allow for a secure way to pass in the token. If not, I'll fall back to #open('token.txt', 'w').write($token)

Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: I meant, a data_source tool like genomespace_push does this automatically: https://github.com/galaxyproject/galaxy/pull/1814/files#diff-164cfd6b30a3542460531fab3994cfb5R10

But a non data_source tool like genomespace_importer needs the exec_before_job hook override to manually generate a json param dict.

Comment (mvdbeek) [reply tools/genomespace/genomespace_exporter.xml]: I think so, does
```
<configfiles>
    <inputs name="inputs" filename="inputs.json" />
<configfiles>
```
work for you ? See https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-configfiles-inputs

Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: I tested this with the exporter. It sends some parameters in, but skips others. For example, of the following:
```
    <inputs>
        <param format="data" name="input1" type="data" label="Send this dataset to GenomeSpace" />
        <!-- If using this tool through bioblend, the genomespace_browser parameter should contain the path to the GenomeSpaceFile + the security token
             separated by a ^ as follows: GenomeSpaceFilePath^Token -->
        <param name="genomespace_browser" type="genomespacefile" label="Choose Target Directory" select_type="FOLDER" />
        <param name="filename" type="text" label="Filename" help="Leave blank to generate automatically" />
    </inputs>
    <outputs>
        <data format="auto" name="output_log" />
    </outputs>
```
Only the following two are sent in:
{"genomespace_browser": "undefined^TOKEN", "filename": "dfsdsfdsfds"}

I also realised that we need the genomespace openid token from user.preferences, so sending in the inputs as is won't help.

I'm now trying the cheetah template #open('token.txt', 'w').write($token) method. But one issue - will that work with Pulsar?

Comment (mvdbeek) [reply tools/genomespace/genomespace_exporter.xml]: Ah, that's unfortunate. I don't think this would be a problem for pulsar, I believe the working directory (where the file ends up) would be staged. You could also try https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-environment-variables-environment-variable instead of writing out the token to a file.

Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: I think the environment variable would be a great fit. The only sensitive piece of information here is the token, and if we can pass that off as an env var, that should be fairly safe. I'll test that now.

File: tools/genomespace/genomespace_exporter.xml
Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: Done

File: tools/genomespace/genomespace_exporter.xml
Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: This worked really well, thanks for point it out.

File: tools/genomespace/genomespace_exporter.xml
Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: Done

File: tools/genomespace/genomespace_importer.xml
Comment (nuwang) [reply tools/genomespace/genomespace_importer.xml]: PYTHONPATH was needed because the gs_importer needs sniff.py and galaxy lib, which fails if it's not explicitly in the python path.

File: tools/genomespace/genomespace_exporter.xml
Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: done

File: tools/genomespace/genomespace_exporter.xml
Comment (nuwang) [reply tools/genomespace/genomespace_exporter.xml]: Removed

**Comment by nuwang** [comment]
@mvdbeek Please take a look now. I think all the issues you've flagged have been fixed. Thanks again for the thorough review.

File: tools/genomespace/genomespace_importer.xml
Comment (mvdbeek) [reply tools/genomespace/genomespace_importer.xml]: OK, in that case you can register the tool id in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/__init__.py#L101, which should preserve the galaxy environment. It should be the same ultimately, but it's nicer to register these tools in a single place.

Comment (nuwang) [reply tools/genomespace/genomespace_importer.xml]: Fixed in commit: https://github.com/galaxyproject/galaxy/pull/1814/commits/07b14b17e7f9fb5dabbbbed49b44e7131ce288ae

**Review by mvdbeek — APPROVED** [review]
Excellent, thank you very much for this contribution @nuwang, and sorry that it took such a long time to review this.

**Comment by nuwang** [comment]
I've rebased so that the client rebuild step is the last commit. I've also enabled edits from maintainers so that the source branch can be edited: https://github.com/gvlproject/galaxy/tree/genomespace_workflows

**Comment by martenson** [comment]
Thank you very much @nuwang for the contribution and @mvdbeek for the detailed review and @jmchilton for the thumbs up.

**Comment by jmchilton** [comment]
Thanks for sticking with this @nuwang - awesome contribution! I feel like I dropped the ball on this and I apologize to everyone, but thanks for picking it up and doing such an awesome job with the review @mvdbeek.

**Pull request merged**
